zio-http/src/main/scala/zio/http/netty/server/ServerInboundHandler.scala

```diff
@@ -28,216 +35,249 @@ private[zio] final case class ServerInboundHandler(
   override def channelRead0(ctx: ChannelHandlerContext, msg: HttpObject): Unit = {
 
-    def addAsyncBodyHandler(async: Body.UnsafeAsync): Unit = {
-      if (contentIsRead) throw new RuntimeException("Content is already read")
-      ctx
-        .channel()
-        .pipeline()
-        .addAfter(Names.HttpRequestHandler, Names.HttpContentHandler, new ServerAsyncBodyHandler(async)): Unit
-      setContentReadAttr(flag = true)
-    }
+    log.debug(s"Message: [${msg.getClass.getName}]")
+    msg match {
+      case jReq: FullHttpRequest =>
+        log.debug(s"FullHttpRequest: [${jReq.method()} ${jReq.uri()}]")
+        val req  = makeZioRequest(ctx, jReq)
+        val exit = http.execute(req)
 
-    def attemptFastWrite(exit: HExit[Any, Throwable, Response], time: service.ServerTime): Boolean = {
-      exit match {
-        case HExit.Success(response) =>
-          response.attribute.encoded match {
-            case Some((oResponse, jResponse: FullHttpResponse)) if hasChanged(response, oResponse) =>
-              val djResponse = jResponse.retainedDuplicate()
-              setServerTime(time, response, djResponse)
-              ctx.writeAndFlush(djResponse, ctx.voidPromise()): Unit
-              // log.debug("Fast write performed")
-              true
-
-            case _ => false
+        val releaseRequest = { () =>
+          if (jReq.refCnt() > 0) {
+            jReq.release(): Unit
           }
-        case _                       => false
-      }
-    }
+        }
 
-    def attemptFullWrite(
-      exit: HExit[Any, Throwable, Response],
-      jRequest: HttpRequest,
-      time: service.ServerTime,
-      runtime: NettyRuntime,
-    ): ZIO[Any, Throwable, Unit] = {
+        if (!attemptImmediateWrite(ctx, exit, time))
+          writeResponse(ctx, env, exit, jReq)(releaseRequest)
+        else
+          releaseRequest()
 
-      for {
-        response <- exit.toZIO.unrefine { case error => Option(error) }.catchAll {
-          case None        => ZIO.succeed(HttpError.NotFound(jRequest.uri()).toResponse)
-          case Some(error) => ZIO.succeed(HttpError.InternalServerError(cause = Some(error)).toResponse)
-        }
-        _        <-
-          if (response.isWebSocket) ZIO.attempt(upgradeToWebSocket(jRequest, response, runtime))
-          else
-            for {
-              jResponse <- response.encode()
-              _         <- ZIO.attemptUnsafe(implicit u => setServerTime(time, response, jResponse))
-              _         <- ZIO.attempt(ctx.writeAndFlush(jResponse))
-              flushed <- if (!jResponse.isInstanceOf[FullHttpResponse]) response.body.write(ctx) else ZIO.succeed(true)
-              _       <- ZIO.attempt(ctx.flush()).when(!flushed)
-            } yield ()
-
-        _ <- ZIO.attemptUnsafe(implicit u => setContentReadAttr(false))
-      } yield log.debug("Full write performed")
-    }
+      case jReq: HttpRequest =>
+        log.debug(s"HttpRequest: [${jReq.method()} ${jReq.uri()}]")
```

Comment:
Debug statement

Reply:
That's been there pre-ziverge.  I can remove it if you feel it is unneccesary.